File: central/src/game-sync.ts
import {digest,readJson,validId} from './security';
type EnvLike={DB:D1Database;RELEASES:R2Bucket};
const now=()=>Math.floor(Date.now()/1000);
const json=(data:unknown,status=200)=>Response.json(data,{status,headers:{'Cache-Control':'no-store'}});
const hash=/^[a-f0-9]{64}$/;
export function safeGamePath(value:unknown):value is string{
 return typeof value==='string'&&value.length<=512&&!/[\\\x00-\x1f:]/.test(value)&&!value.startsWith('/')&&value.split('/').every(p=>p!=='.'&&p!=='..'&&p!=='')&&!/^\d+(\/|$)|^start_fs25_|(^|\/)(savegame\d*|mods|\.env|game\.xml|dedicatedServer\.xml|log\.txt|cert\.pem|pk\.pem)(\/|$)/i.test(value);
}
export function validateManifest(body:any){
 if(!body||!/^\d+(\.\d+){2,3}$/.test(body.version)||!Array.isArray(body.files)||!body.files.length||body.files.length>20000)throw new Error('Invalid game manifest');
 const seen=new Set();let total=0,chunks=0;
 for(const f of body.files){
  if(!safeGamePath(f.path)||seen.has(f.path)||!Number.isSafeInteger(f.size)||f.size<0||f.size>100*1024**3||!Array.isArray(f.chunks)||f.chunks.length!==Math.ceil(f.size/(4*1024**2))||f.chunks.some((c:unknown)=>typeof c!=='string'||!hash.test(c)))throw new Error('Invalid game file');
  seen.add(f.path);total+=f.size;chunks+=f.chunks.length;
 }
 if(total>200*1024**3||chunks>100000||!seen.has('x64/FarmingSimulator2025Game.exe')||!seen.has('dedicatedServer.exe')||!seen.has('FarmingSimulator2025.exe')||!seen.has('VERSION'))throw new Error('Incomplete or oversized game installation');
 return total;
}
async function queue(env:EnvLike,node:string,kind:string,release:string|null){
 const t=now();await env.DB.prepare("INSERT OR IGNORE INTO game_jobs(id,node_id,kind,release_id,created,updated) SELECT ?,id,?,?,?,? FROM nodes WHERE id=? AND enabled=1").bind(crypto.randomUUID(),kind,release,t,t,node).run();
}
export async function gameNode(request:Request,env:EnvLike){
 const url=new URL(request.url),op=url.pathname.slice('/api/game-node/'.length),node=request.headers.get('X-Node-ID')||'',bearer=request.headers.get('Authorization')||'';
 if(!validId(node)||!/^Bearer [a-f0-9]{64}$/.test(bearer)||!await env.DB.prepare('SELECT id FROM nodes WHERE id=? AND token_hash=? AND enabled=1').bind(node,await digest(bearer.slice(7))).first())return json({error:'Unauthorized node'},401);
 if(op==='poll'&&request.method==='POST'){
  const b=await readJson(request);if(!['ready','empty','unknown','error'].includes(String(b.status))||typeof b.detail!=='string'||b.detail.length>500||b.version!==null&&(typeof b.version!=='string'||!/^\d+(\.\d+){2,3}$/.test(b.version))||b.fingerprint!==null&&(typeof b.fingerprint!=='string'||!hash.test(b.fingerprint)))return json({error:'Invalid inventory'},422);
  await env.DB.prepare('INSERT INTO game_inventory(node_id,version,fingerprint,status,detail,updated) VALUES(?,?,?,?,?,?) ON CONFLICT(node_id) DO UPDATE SET version=excluded.version,fingerprint=excluded.fingerprint,status=excluded.status,detail=excluded.detail,updated=excluded.updated').bind(node,b.version,b.fingerprint,b.status,b.detail,now()).run();
  const policy=await env.DB.prepare('SELECT * FROM game_policy WHERE id=1').first<any>();
  if(b.status==='empty'&&policy?.auto_bootstrap&&policy.release_id)await queue(env,node,'sync',policy.release_id);
  const job=await env.DB.prepare("SELECT * FROM game_jobs WHERE node_id=? AND status IN ('queued','running','waiting') ORDER BY created LIMIT 1").bind(node).first();if(job){const selections=(await env.DB.prepare('SELECT instance_id,packages FROM game_dlc_settings WHERE node_id=?').bind(node).all<any>()).results;const reported=await env.DB.prepare('SELECT snapshot FROM nodes WHERE id=?').bind(node).first<any>();(job as any).dlc_settings=(JSON.parse(reported?.snapshot||'{}')?.servers||[]).map((s:any)=>({instance:s.instance_id,packages:JSON.parse(selections.find(x=>x.instance_id===s.instance_id)?.packages||'[]')}));(job as any).dlcs=[...new Set(selections.flatMap(s=>JSON.parse(s.packages)))];}return json({job});
 }
 const jobId=url.searchParams.get('job')||'';
 const job=await env.DB.prepare("SELECT * FROM game_jobs WHERE id=? AND node_id=? AND status IN ('queued','running','waiting')").bind(jobId,node).first<any>();
 if(!job)return json({error:'Active job not found'},404);
 if(op==='progress'&&request.method==='POST'){
  const b=await readJson(request);if(!['running','waiting','failed','succeeded'].includes(String(b.status))||typeof b.detail!=='string'||b.detail.length>500||![b.done,b.total].every(v=>typeof v==='number'&&Number.isSafeInteger(v)&&v>=0&&v<=200*1024**3)||Number(b.done)>Number(b.total)||job.kind==='publish'&&b.status==='succeeded')return json({error:'Invalid progress'},422);
  await env.DB.prepare('UPDATE game_jobs SET status=?,detail=?,done_bytes=?,total_bytes=?,updated=? WHERE id=?').bind(b.status,b.detail,b.done,b.total,now(),jobId).run();return json({ok:true});
 }
 const chunk=url.searchParams.get('hash')||'';
 if(op==='chunk'&&job.kind==='publish'&&hash.test(chunk)&&request.method==='PUT'){
  const reader=request.body?.getReader();if(!reader)return json({error:'Missing chunk'},422);let size=0;const parts:Uint8Array[]=[];
  try{while(true){const r=await reader.read();if(r.done)break;size+=r.value.length;if(size>4*1024**2)return json({error:'Chunk too large'},413);parts.push(r.value);}}finally{await reader.cancel();}
  const bytes=new Uint8Array(size);let offset=0;for(const p of parts){bytes.set(p,offset);offset+=p.length;}
  const actual=Array.from(new Uint8Array(await crypto.subtle.digest('SHA-256',bytes)),b=>b.toString(16).padStart(2,'0')).join('');if(actual!==chunk)return json({error:'Checksum mismatch'},422);
  await env.RELEASES.put('game-library/'+node+'/chunks/'+chunk,bytes);return json({ok:true});
 }
 if(op==='publish'&&job.kind==='publish'&&request.method==='POST'){
  const b=await readJson(request,8*1024**2);let size:number;try{size=validateManifest(b);}catch{return json({error:'Invalid game manifest'},422);}
  const raw=JSON.stringify(b),fingerprint=await digest(JSON.stringify({version:b.version,files:(b.files as any[]).filter(f=>!f.path.startsWith('__dlc__/'))})),key='game-library/'+node+'/manifests/'+await digest(raw);
  await env.RELEASES.put(key,raw);
  await env.DB.batch([env.DB.prepare('INSERT OR IGNORE INTO game_releases(id,source_node,version,fingerprint,manifest_key,bytes,dlcs,created) VALUES(?,?,?,?,?,?,?,?)').bind(jobId,node,b.version,fingerprint,key,size,JSON.stringify([...new Set((b.files as any[]).map(f=>/^__dlc__\/FarmingSimulator25_([^_]+)_/.exec(f.path)?.[1]).filter(Boolean))]),now()),env.DB.prepare("UPDATE game_jobs SET status='succeeded',done_bytes=?,total_bytes=?,detail='Game version published. Approve it on Game status.',updated=? WHERE id=?").bind(size,size,now(),jobId)]);return json({ok:true});
 }
 if(job.kind==='sync'&&request.method==='GET'){
  const release=await env.DB.prepare('SELECT * FROM game_releases WHERE id=?').bind(job.release_id).first<any>();if(!release)return json({error:'Release missing'},404);
  if(op==='manifest'){const object=await env.RELEASES.get(release.manifest_key);return object?new Response(object.body,{headers:{'Content-Type':'application/json','Cache-Control':'no-store'}}):json({error:'Manifest missing'},404);}
  if(op==='chunk'&&hash.test(chunk)){
   // Downloads are restricted to the assigned release's source namespace.
   const object=await env.RELEASES.get('game-library/'+release.source_node+'/chunks/'+chunk);return object?new Response(object.body,{headers:{'Content-Type':'application/octet-stream','Cache-Control':'no-store'}}):json({error:'Chunk missing'},404);
  }
 }
 return json({error:'Unknown game sync operation'},404);
}
export async function gameAdmin(request:Request,env:EnvLike){
 if(request.method==='GET')return json({nodes:(await env.DB.prepare('SELECT n.id,n.name,n.enabled,n.last_seen,n.snapshot,i.version,i.fingerprint,i.status,i.detail,i.updated FROM nodes n LEFT JOIN game_inventory i ON i.node_id=n.id ORDER BY n.name').all()).results,releases:(await env.DB.prepare('SELECT * FROM game_releases ORDER BY created DESC LIMIT 100').all()).results,jobs:(await env.DB.prepare('SELECT * FROM game_jobs ORDER BY created DESC LIMIT 100').all()).results,dlcSettings:(await env.DB.prepare('SELECT * FROM game_dlc_settings').all()).results,policy:await env.DB.prepare('SELECT * FROM game_policy WHERE id=1').first()});
 if(request.method!=='POST')return json({error:'Method not allowed'},405);
 const b=await readJson(request);
 if(b.action==='dlcs'&&validId(b.node)&&typeof b.instance==='string'&&Array.isArray(b.packages)&&b.packages.length<=100&&b.packages.every(p=>typeof p==='string'&&/^[A-Za-z0-9-]{1,100}$/.test(p))){
  const n=await env.DB.prepare('SELECT snapshot FROM nodes WHERE id=? AND enabled=1').bind(b.node).first<any>();if(!n||!JSON.parse(n.snapshot||'{}')?.servers?.some((s:any)=>s.instance_id===b.instance))return json({error:'Unknown server'},422);
  await env.DB.prepare('INSERT INTO game_dlc_settings(node_id,instance_id,packages) VALUES(?,?,?) ON CONFLICT(node_id,instance_id) DO UPDATE SET packages=excluded.packages').bind(b.node,b.instance,JSON.stringify([...new Set(b.packages)])).run();return json({ok:true});
 }
 if(b.action==='publish'&&validId(b.node)){await queue(env,b.node,'publish',null);return json({ok:true});}
 if(b.action==='approve'&&typeof b.release==='string'&&await env.DB.prepare('SELECT id FROM game_releases WHERE id=?').bind(b.release).first()){await env.DB.prepare('UPDATE game_policy SET release_id=? WHERE id=1').bind(b.release).run();return json({ok:true});}
 if(b.action==='bootstrap'&&typeof b.enabled==='boolean'){await env.DB.prepare('UPDATE game_policy SET auto_bootstrap=? WHERE id=1').bind(b.enabled?1:0).run();return json({ok:true});}
 if(b.action==='sync'){
  const policy=await env.DB.prepare('SELECT release_id FROM game_policy WHERE id=1').first<any>();if(!policy?.release_id)return json({error:'Approve a published version first'},409);
  const release=await env.DB.prepare('SELECT fingerprint,source_node FROM game_releases WHERE id=?').bind(policy.release_id).first<any>();
  const targets=await env.DB.prepare('SELECT n.id,i.fingerprint FROM nodes n LEFT JOIN game_inventory i ON i.node_id=n.id WHERE n.enabled=1').all<any>();
  for(const n of targets.results)if(b.node===n.id||b.node==='all'&&(n.id!==release.source_node||n.fingerprint!==release.fingerprint))await queue(env,n.id,'sync',policy.release_id);return json({ok:true});
 }
 return json({error:'Invalid game status action'},422);
}

File: scripts/game-sync.py
#!/usr/bin/env python3
"""Private, resumable game library worker. Runs on a node, never in a browser."""
import fcntl
import hashlib
import json
import os
from pathlib import Path, PurePosixPath
import re
import shutil
import subprocess
import time
import urllib.request
import urllib.error

CHUNK = 4 * 1024 * 1024
STATE = Path('/var/lib/farmservers-game-sync')
SITE = 'https://farmservers.sargentweb.com/api/game-node/'
BLOCKED = {'game.xml', 'dedicatedserver.xml', 'log.txt', 'cert.pem', 'pk.pem', '.env'}


def safe_path(name):
    p = PurePosixPath(name)
    return (isinstance(name, str) and len(name) <= 512 and not re.search(r'[\\\x00-\x1f:]', name)
            and not p.is_absolute() and name == p.as_posix() and all(part not in ('.', '..', '') for part in name.split('/'))
            and not re.match(r'^\d+(/|$)|^start_fs25_', name, re.I)
            and not any(part.lower() in BLOCKED or re.fullmatch(r'savegame\d*|mods', part, re.I) for part in p.parts))


def target(root, name):
    if not safe_path(name):
        raise ValueError('Unsafe game file path')
    result = root / name
    for path in [root, result, *result.parents]:
        if path.is_symlink():
            raise ValueError('Symlinks are not supported in game sync')
        if path == root:
            break
    return result


def save(path, value):
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.with_suffix('.tmp')
    with tmp.open('w') as stream:
        json.dump(value, stream, separators=(',', ':'))
        stream.flush()
        os.fsync(stream.fileno())
    tmp.replace(path)


class NoRedirect(urllib.request.HTTPRedirectHandler):
    def redirect_request(self, *args, **kwargs):
        raise ValueError('Redirect refused')


def request(config, operation, body=None, raw=False, method=None, limit=8*1024*1024):
    data = body if raw else json.dumps(body, separators=(',', ':')).encode() if body is not None else None
    req = urllib.request.Request(SITE+operation, data=data, method=method, headers={
        'Authorization': 'Bearer '+config['token'], 'X-Node-ID': config['node'],
        # Cloudflare bot protection blocks Python's default User-Agent with HTTP 403.
        'User-Agent': 'Farmservers-Node/1.0 (+https://farmservers.sargentweb.com)',
        'Content-Type': 'application/octet-stream' if raw else 'application/json'})
    with urllib.request.build_opener(NoRedirect()).open(req, timeout=120) as response:
        data = response.read(limit+1)
    if len(data)>limit:
        raise ValueError('Response too large')
    return data if raw else json.loads(data)


def version(root):
    path = target(root, 'VERSION')
    if not path.is_file():
        return None
    with path.open() as stream:value = stream.read(100).strip().lstrip('v')
    return value if re.fullmatch(r'\d+(\.\d+){2,3}', value) else None


def manifest(root, dlc_root):
    game_version = version(root)
    if not game_version:
        raise ValueError('Game VERSION file is missing or invalid; complete the game installation first')
    cache_path=STATE/'scan-cache.json'
    cache=json.loads(cache_path.read_text()) if cache_path.exists() else {}
    next_cache={}
    files = []
    total_size=0
    roots = [(root, '')]
    if dlc_root.is_dir():
        roots.append((dlc_root, '__dlc__/'))
    for base, prefix in roots:
        for directory, dirs, names in os.walk(base, followlinks=False):
            dirs[:] = sorted(d for d in dirs if not d.startswith('.') and not (Path(directory)/d).is_symlink() and safe_path((Path(directory)/d).relative_to(base).as_posix()))
            for name in sorted(names):
                file = Path(directory)/name
                relative = file.relative_to(base).as_posix()
                if file.is_symlink() or not safe_path(relative) or name.startswith('.') or prefix and not re.fullmatch(r'FarmingSimulator25_[A-Za-z0-9-]+_[A-Za-z0-9_.-]+\.exe', relative):
                    continue
                chunks = []
                stat = file.stat()
                total_size+=stat.st_size
                if stat.st_size>100*1024**3 or total_size>200*1024**3:raise ValueError('Game library exceeds size limits')
                cache_key=str(file)
                cached=cache.get(cache_key,{})
                if cached.get('mtime')==stat.st_mtime_ns and cached.get('size')==stat.st_size:
                    chunks=cached['chunks']
                else:
                    with file.open('rb') as stream:
                        while data := stream.read(CHUNK):
                            chunks.append(hashlib.sha256(data).hexdigest())
                next_cache[cache_key]={'mtime':stat.st_mtime_ns,'size':stat.st_size,'chunks':chunks}
                if file.stat().st_mtime_ns != stat.st_mtime_ns or file.stat().st_size != stat.st_size:
                    raise ValueError('Game files changed during scan; retry after the update finishes')
                files.append({'path': prefix+relative, 'size':stat.st_size, 'chunks':chunks})
                if len(files)>20000:
                    raise ValueError('Too many game files')
    files.sort(key=lambda f:f['path'])
    if not {'x64/FarmingSimulator2025Game.exe','dedicatedServer.exe','FarmingSimulator2025.exe','VERSION'}.issubset({f['path'] for f in files}):
        raise ValueError('Game installation is incomplete')
    save(cache_path,next_cache)
    return {'version':game_version,'files':files}


def fingerprint(value):
    base={'version':value['version'],'files':[f for f in value['files'] if not f['path'].startswith('__dlc__/')]}
    return hashlib.sha256(json.dumps(base,separators=(',', ':'),ensure_ascii=False).encode()).hexdigest()


def uses_game(row,root):
    # The control agent also mounts /opt and shared storage; it must stay online.
    if row.get('Name')=='/fsg-panel-agent' and row.get('Config',{}).get('Labels',{}).get('com.docker.compose.service')=='agent':return False
    return any((Path(m['Source'])==root or root.is_relative_to(Path(m['Source'])) or Path(m['Source']).is_relative_to(root)) and m.get('Destination')=='/opt/fs25/game' for m in row.get('Mounts',[]) if m.get('Type')=='bind')


def running(root):
    ids=subprocess.run(['docker','ps','-q'],check=True,capture_output=True,text=True,timeout=30).stdout.split()
    if not ids:
        return False
    rows=json.loads(subprocess.run(['docker','inspect',*ids],check=True,capture_output=True,text=True,timeout=30).stdout)
    return any(uses_game(row,root) for row in rows)


def dlc_runtime_ready(root):
    ids=subprocess.run(['docker','ps','-aq'],check=True,capture_output=True,text=True,timeout=30).stdout.split()
    if not ids:return True
    rows=json.loads(subprocess.run(['docker','inspect',*ids],check=True,capture_output=True,text=True,timeout=30).stdout)
    for row in rows:
        if uses_game(row,root) and row.get('Config',{}).get('Labels',{}).get('com.farmservers.dlc-selection')!='1':return False
    return True


def paths(config):
    script="require '/var/www/html/src/bootstrap.php'; $h=local_host_record(); echo json_encode(['game'=>$h['shared_game_path'],'dlc'=>$h['shared_dlc_path'],'instances'=>env_value('INSTANCE_BASE_PATH','/opt/fsg-panel/instances')]);"
    result=subprocess.run(['docker','compose','-p',config['project'],'exec','-T','web','php','-r',script],cwd=config['root'],check=True,capture_output=True,text=True,timeout=60)
    values=json.loads(result.stdout)
    game=Path(values['game'])/'Farming Simulator 2025'
    dlc=Path(values['dlc'])
    if not game.is_absolute() or not dlc.is_absolute() or len(dlc.parts)<3 or any(p.is_symlink() for p in [game,dlc,*game.parents,*dlc.parents]):
        raise ValueError('Unsafe shared storage paths')
    return game,dlc,Path(values['instances'])


_last_progress={}

def progress(config,job,status,done,total,detail):
    stamp=time.monotonic()
    if status=='running' and done<total and stamp-_last_progress.get(job['id'],0)<1:return
    _last_progress[job['id']]=stamp
    request(config,'progress?job='+job['id'],{'status':status,'done':done,'total':total,'detail':detail[:500]})


def publish(config,job,root,dlc):
    data=manifest(root,dlc)
    total=sum(f['size'] for f in data['files']);done=0
    journal=STATE/(job['id']+'-uploaded.json')
    uploaded=set(json.loads(journal.read_text())) if journal.exists() else set()
    for file in data['files']:
        name=file['path'];base=dlc if name.startswith('__dlc__/') else root
        path=target(base,name.removeprefix('__dlc__/'))
        with path.open('rb') as stream:
            for chunk in file['chunks']:
                raw=stream.read(CHUNK)
                if hashlib.sha256(raw).hexdigest()!=chunk:
                    raise ValueError('Source files changed; publish again after the game update completes')
                if chunk not in uploaded:
                    request(config,'chunk?job='+job['id']+'&hash='+chunk,raw,raw=True,method='PUT')
                    uploaded.add(chunk);save(journal,sorted(uploaded))
                done+=len(raw)
                progress(config,job,'running',done,total,'Publishing '+name)
    if manifest(root,dlc)!=data:
        raise ValueError('Source installation changed during publication')
    request(config,'publish?job='+job['id'],data)


def install(config,job,root,dlc,instances):
    if job.get('status')=='waiting' and running(root):
        progress(config,job,'waiting',job.get('done_bytes',0),job.get('total_bytes',0),'Files downloaded. Stop all game containers on this node to apply.');return
    completed=STATE/(job['id']+'-complete.json')
    if completed.exists():
        info=json.loads(completed.read_text());progress(config,job,'succeeded',info['total'],info['total'],'Game files ready.');return
    data=request(config,'manifest?job='+job['id'])
    if not isinstance(data.get('files'),list) or len(data['files'])>20000:
        raise ValueError('Invalid manifest')
    stage=STATE/job['id'];stage.mkdir(exist_ok=True)
    contents=stage/'files';contents.mkdir(exist_ok=True)
    # The library may include optional DLC installers. Only selected packages are copied.
    selected=set(job.get('dlcs',[]))
    files=[f for f in data['files'] if not f['path'].startswith('__dlc__/') or re.match(r'__dlc__/FarmingSimulator25_([^_]+)_',f['path']).group(1) in selected]
    if not all(safe_path(f.get('path','')) for f in files):raise ValueError('Unsafe manifest')
    total=sum(f['size'] for f in files);done=0
    missing=sum(max(0,f['size']-(target(contents,f['path']).stat().st_size if target(contents,f['path']).is_file() else 0)) for f in files)
    if shutil.disk_usage(stage).free<missing+total:
        raise ValueError('Not enough free disk space for staging and rollback backup')
    for file in files:
        name=file['path'];dest=target(contents,name);dest.parent.mkdir(parents=True,exist_ok=True)
        if not isinstance(file['size'],int) or file['size']<0 or len(file['chunks'])!=(file['size']+CHUNK-1)//CHUNK:
            raise ValueError('Invalid file size')
        with dest.open('a+b') as output:
            for index,chunk in enumerate(file['chunks']):
                if not re.fullmatch('[a-f0-9]{64}',chunk):
                    raise ValueError('Invalid checksum')
                offset=index*CHUNK;size=min(CHUNK,file['size']-offset)
                output.seek(offset);raw=output.read(size)
                if hashlib.sha256(raw).hexdigest()!=chunk:
                    raw=request(config,'chunk?job='+job['id']+'&hash='+chunk,raw=True,limit=CHUNK)
                    if len(raw)!=size or hashlib.sha256(raw).hexdigest()!=chunk:
                        raise ValueError('Downloaded file failed verification')
                    # a+b always appends; reopen for bounded, resumable random writes.
                    with dest.open('r+b') as write:
                        write.seek(offset);write.write(raw);write.flush();os.fsync(write.fileno())
                done+=size;progress(config,job,'running',done,total,'Downloading '+name)
            output.truncate(file['size'])
    if selected and not dlc_runtime_ready(root):
        progress(config,job,'waiting',done,total,'Rebuild/recreate game containers with the latest runtime before applying per-server DLC selections.');return
    if running(root):
        progress(config,job,'waiting',done,total,'Files downloaded. Stop all game containers on this node to apply.');return
    root.mkdir(parents=True,exist_ok=True);dlc.mkdir(parents=True,exist_ok=True)
    root.chmod(0o755);dlc.chmod(0o755)
    env_text=(Path(config['root'])/'.env').read_text() if config.get('root') and (Path(config['root'])/'.env').is_file() else ''
    uid=int(re.search(r'^PUID=(\d+)$',env_text,re.M).group(1)) if re.search(r'^PUID=(\d+)$',env_text,re.M) else 1000
    gid=int(re.search(r'^PGID=(\d+)$',env_text,re.M).group(1)) if re.search(r'^PGID=(\d+)$',env_text,re.M) else 1000
    if hasattr(os,'chown'):
        os.chown(root,uid,gid);os.chown(dlc,uid,gid)
    apply_lock=(root.parent/'.farmservers-sync.lock').open('a')
    try:
        fcntl.flock(apply_lock,fcntl.LOCK_EX|fcntl.LOCK_NB)
    except BlockingIOError:
        apply_lock.close();progress(config,job,'waiting',done,total,'Waiting for a server operation to finish.');return
    lock=target(root,'.farmservers-sync-in-progress')
    lock.write_text(job['id'])
    try:
        if running(root):
            progress(config,job,'waiting',done,total,'Waiting for game containers to stop.');return
        backup=stage/'backup';journal=stage/'apply.json'
        if journal.exists():
            # Never continue a partially applied installation without restoring its old files.
            rollback(root,dlc,backup,json.loads(journal.read_text()));journal.unlink()
        entries=[]
        for file in files:
            name=file['path'];base=dlc if name.startswith('__dlc__/') else root;dest=target(base,name.removeprefix('__dlc__/'))
            entries.append({'path':name,'existed':dest.exists()})
            if dest.exists():
                old=target(backup,name);old.parent.mkdir(parents=True,exist_ok=True);shutil.copy2(dest,old)
        save(journal,entries)
        try:
            for file in files:
                name=file['path'];base=dlc if name.startswith('__dlc__/') else root;dest=target(base,name.removeprefix('__dlc__/'));dest.parent.mkdir(parents=True,exist_ok=True)
                for folder in [dest.parent,*dest.parent.parents]:
                    folder.chmod(0o755)
                    if hasattr(os,'chown'):os.chown(folder,uid,gid)
                    if folder==base:break
                tmp=dest.with_name(dest.name+'.farmservers-new');shutil.copyfile(target(contents,name),tmp);tmp.chmod(0o644)
                if hasattr(os,'chown'):os.chown(tmp,uid,gid)
                with tmp.open('r+b') as stream:os.fsync(stream.fileno())
                tmp.replace(dest)
            # TLS credentials are unique to this node, never copied from the source.
            if not (root/'cert.pem').exists() or not (root/'pk.pem').exists():
                subprocess.run(['openssl','req','-x509','-newkey','rsa:2048','-nodes','-keyout',str(root/'pk.pem'),'-out',str(root/'cert.pem'),'-days','365','-subj','/CN=localhost'],check=True,capture_output=True,timeout=60)
                (root/'pk.pem').chmod(0o644)
            (root/'.game-install-complete').touch()
            for setting in job.get('dlc_settings',[]):
                if not re.fullmatch(r'[A-Za-z0-9_-]{1,64}',setting['instance']):raise ValueError('Invalid DLC target')
                instance=target(instances,setting['instance'])
                if not instance.is_dir():continue
                selection=target(instance,'data/config/FarmingSimulator2025/.farmservers-enabled-dlcs')
                missing=[];parent=selection.parent
                while not parent.exists():missing.append(parent);parent=parent.parent
                selection.parent.mkdir(parents=True,exist_ok=True)
                for folder in missing:
                    folder.chmod(0o755)
                    if hasattr(os,'chown'):os.chown(folder,uid,gid)
                selection.write_text('\n'.join(setting['packages'])+'\n');selection.chmod(0o644)
            journal.unlink()
        except Exception:
            rollback(root,dlc,backup,entries);journal.unlink();raise
        save(completed,{'total':total})
        progress(config,job,'succeeded',total,total,'Game files ready. Create a server and complete its own license activation.')
    finally:
        if not (stage/'apply.json').exists():lock.unlink(missing_ok=True)
        apply_lock.close()


def rollback(root,dlc,backup,entries):
    for entry in entries:
        name=entry['path'];base=dlc if name.startswith('__dlc__/') else root;dest=target(base,name.removeprefix('__dlc__/'))
        if entry['existed']:
            shutil.copy2(target(backup,name),dest)
        else:
            dest.unlink(missing_ok=True)


def main():
    os.umask(0o077);STATE.mkdir(parents=True,exist_ok=True)
    with (STATE/'lock').open('w') as lock:
        fcntl.flock(lock,fcntl.LOCK_EX|fcntl.LOCK_NB)
        config=json.loads(Path('/etc/farmservers/updater.json').read_text())
        root,dlc,instances=paths(config)
        marker=root/'.farmservers-sync-in-progress'
        if marker.exists():
            recovery_id=marker.read_text().strip()
            if not re.fullmatch('[a-f0-9-]{36}',recovery_id):raise ValueError('Invalid recovery marker')
            recovery=STATE/recovery_id/'apply.json'
            with (root.parent/'.farmservers-sync.lock').open('a') as recovery_lock:
                fcntl.flock(recovery_lock,fcntl.LOCK_EX|fcntl.LOCK_NB)
                if running(root):raise ValueError('Stop game containers before sync recovery')
                if recovery.exists():
                    rollback(root,dlc,STATE/recovery_id/'backup',json.loads(recovery.read_text()));recovery.unlink()
                marker.unlink()
        inventory={'version':None,'fingerprint':None,'status':'empty','detail':'No game installation yet.'}
        if root.exists() and any(p.name!='.farmservers-sync-in-progress' for p in root.iterdir()):
            try:
                data=manifest(root,dlc);inventory={'version':data['version'],'fingerprint':fingerprint(data),'status':'ready','detail':'Game files detected.'}
            except Exception as error:
                inventory.update(status='unknown',version=version(root),detail=str(error)[:500])
        job=request(config,'poll',inventory)['job']
        if job:
            if not re.fullmatch('[a-f0-9-]{36}',job['id']):raise ValueError('Invalid job ID')
            try:
                if job['kind']=='publish':publish(config,job,root,dlc)
                else:install(config,job,root,dlc,instances)
            except urllib.error.HTTPError as error:
                if error.code in (408,429) or error.code>=500:raise
                progress(config,job,'failed',0,0,'Game library request rejected (HTTP '+str(error.code)+'). Check node software and retry.')
            except (urllib.error.URLError,TimeoutError):
                raise  # Keep the active job resumable on the next timer tick.
            except Exception as error:
                progress(config,job,'failed',0,0,str(error))


if __name__=='__main__':
    main()
